src/python/pants/source/payload_fields.py

```diff
@@ -54,12 +61,14 @@ def source_paths(self):
 
   @property
   def address(self):
-    """Returns the address this sources field refers to (used by some derived classses)"""
+    """Returns the address this sources field refers to (used by some derived classes)"""
     return self._ref_address
 
   def has_sources(self, extension=None):
     if not self.source_paths:
       return False
+    if not extension:
```

Comment:
Hm... "True if no extension" is an odd case. AFAICT, this method is only used in Target... would probably be good to inline the logic there so that it is more obvious that it is completely overridden by `Resources`. Would that also make this check unnecessary?